drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,659 @@
+
+	if (!device_is_ready(config->sdhi_on_gpio.port)) {
+		LOG_ERR("Card is not ready.");
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&config->sdhi_on_gpio, GPIO_OUTPUT_HIGH);
+	if (ret) {
+		LOG_ERR("Error %d: Card configuration failed.", ret);
+		return ret;
+	}
+
+	k_msleep(250);
+	return ret;
+}
+
+/**
+ * Logs detailed SDIO error types using Zephyr's logging subsystem.
+ *
+ * This helper function queries the error status of an SDIO operation
+ * and reports specific error types using `LOG_ERR()`.
+ *
+ * @param hsd Pointer to the SDIO handle.
+ */
+void sdhc_stm32_log_err_type(SDIO_HandleTypeDef *hsd)
+{
+	uint32_t error_code = HAL_SDIO_GetError(hsd);
+
+	if (error_code & HAL_SDIO_ERROR_TIMEOUT) {
+		LOG_ERR("SDIO Timeout Error\n");
+	}
+
+	if (error_code & HAL_SDIO_ERROR_DATA_TIMEOUT) {
+		LOG_ERR("SDIO Data Timeout Error\n");
+	}
+
+	if (error_code & HAL_SDIO_ERROR_DATA_CRC_FAIL) {
+		LOG_ERR("SDIO Data CRC Error\n");
+	}
+
+	if (error_code & HAL_SDIO_ERROR_TX_UNDERRUN) {
+		LOG_ERR("SDIO FIFO Transmit Underrun Error\n");
+	}
+
+	if (error_code & HAL_SDIO_ERROR_RX_OVERRUN) {
+		LOG_ERR("SDIO FIFO Receive Overrun Error\n");
+	}
+
+	if (error_code & HAL_SDIO_ERROR_INVALID_CALLBACK) {
+		LOG_ERR("SDIO Invalid Callback Error\n");
+	}
+
+	if (error_code & SDMMC_ERROR_ADDR_MISALIGNED) {
+		LOG_ERR("SDIO Misaligned address Error\n");
+	}
+
+	if (error_code & SDMMC_ERROR_WRITE_PROT_VIOLATION) {
+		LOG_ERR("Attempt to program a write protect block\n");
+	}
+
+	if (error_code & SDMMC_ERROR_ILLEGAL_CMD) {
+		LOG_ERR("Command is not legal for the card state\n");
+	}
+
+	hsd->ErrorCode = HAL_SDIO_ERROR_NONE;
+}
+
+/**
+ * Initializes the SDIO peripheral with the configuration specified.
+ *
+ * This includes deinitializing any previous configuration, and applying
+ * parameters like clock edge, power saving, clock divider, hardware
+ * flow control and bus width.
+ *
+ * @param dev Pointer to the device structure for the SDIO peripheral.
+ *
+ * @return HAL_OK on success, HAL_ERROR on failure.
+ */
+int sdhc_stm32_sd_init(const struct device *dev)
+{
+	struct sdhc_stm32_data *data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+	SDIO_HandleTypeDef *hsd = &data->hsd;
+
+	if (HAL_SDIO_DeInit(hsd) != HAL_OK) {
+		LOG_ERR("Failed to de-initialize the SDIO device\n");
+		return HAL_ERROR;
+	}
+
+	hsd->Init.ClockEdge = SDMMC_CLOCK_EDGE_FALLING;
+	hsd->Init.ClockPowerSave = SDMMC_CLOCK_POWER_SAVE_DISABLE;
+	hsd->Init.ClockDiv = config->clk_div;
+
+	if (config->hw_flow_control) {
+		hsd->Init.HardwareFlowControl = SDMMC_HARDWARE_FLOW_CONTROL_ENABLE;
+	} else {
+		hsd->Init.HardwareFlowControl = SDMMC_HARDWARE_FLOW_CONTROL_DISABLE;
+	}
+
+	if (data->host_io.bus_width == 4) {
+		hsd->Init.BusWide = SDMMC_BUS_WIDE_4B;
+	} else if (data->host_io.bus_width == 8) {
+		hsd->Init.BusWide = SDMMC_BUS_WIDE_8B;
+	} else {
+		hsd->Init.BusWide = SDMMC_BUS_WIDE_1B;
+	}
+
+	return HAL_SDIO_Init(hsd);
```

Comment:
This function is slightly too SDIO specific- the HAL init code is writing to SDIO card registers. The code there related to the 74 init clocks/SDHC register configuration is fine, but we can't have SDIO-specific code within the SDHC init sequence that won't work with SDMMC cards